### PR TITLE
Enable strict_equality for mypy #3579

### PR DESCRIPTION
**Pull request by wattlebirdaz**
## Motivation

Refs #3579

## Description of the changes

Enable strict_equality option for mypy

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@toshihikoyanase @c-bata Could you review this PR?

**Review by c-bata — APPROVED** [review]
LGTM! Thank you for your contribution :tada:

File: tests/storages_tests/test_heartbeat.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def test_fail_stale_trials_with_optimize(storage_mode: str) -> None:
         with patch("optuna.study._optimize.Thread", _TestableThread):
             study1.optimize(lambda _: 1.0, n_trials=1)
 
-        assert study1.trials[0].state is TrialState.FAIL
+        assert study1.trials[0].state is TrialState.FAIL  # type: ignore [comparison-overlap]
```
Comment (toshihikoyanase) [inline tests/storages_tests/test_heartbeat.py:52]: [Notes]  `mypy` inferred that `study1.trials[0].state` was `Literal[TrialState.RUNNING]` due to the assertion at L45.

File: tests/trial_tests/test_trial.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def test_check_distribution_suggest_uniform(storage_mode: str) -> None:
         study = create_study(storage=storage, sampler=sampler)
         trial = Trial(study, study._storage.create_new_trial(study._study_id))
 
-        with pytest.warns(None) as record:
+        with warnings.catch_warnings(record=True) as record:
```
Comment (toshihikoyanase) [inline tests/trial_tests/test_trial.py:76]: [Question] This might not be related to `strict_equality` since it causes no errors when I revert the change. Could you tell me why you introduced this change?

**Review by toshihikoyanase — COMMENTED** [review]
I'm sorry for the delayed response. I have a question on the change in the test.

File: tests/trial_tests/test_trial.py (hunk 2 above)
Comment (wattlebirdaz) [reply tests/trial_tests/test_trial.py:76]: I reverted the change in `test/trial_tests/test_trial.py L.76` and re-run mypy. I got the following results. 

```sh
optuna$ mypy --version
mypy 0.961 (compiled: yes)
optuna$ mypy tests
tests/trial_tests/test_trial.py:76: error: No overload variant of "warns" matches argument type "None"
tests/trial_tests/test_trial.py:76: note: Possible overload variants:
tests/trial_tests/test_trial.py:76: note:     def warns(expected_warning: Union[Type[Warning], Tuple[Type[Warning], ...]] = ..., *, match: Union[str, Pattern[str], None] = ...) -> WarningsChecker
tests/trial_tests/test_trial.py:76: note:     def [T] warns(expected_warning: Union[Type[Warning], Tuple[Type[Warning], ...]], func: Callable[..., T], *args: Any, **kwargs: Any) -> T
```

I fixed this referring to [this](https://github.com/scikit-learn/scikit-learn/issues/22572).

Comment (toshihikoyanase) [reply tests/trial_tests/test_trial.py:76]: Thank you for your confirmation. I reproduced the error. It was due to  `pytest` and the `checks` CI job cannot detect it because `pytest` is not installed in the job. https://github.com/optuna/optuna/pull/3580 will resolve it, so could you revert the changes in `test_trial.py`, please?

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update! LGTM.
